cdc/owner.go

```diff
@@ -1103,3 +1108,194 @@ func (o *ownerImpl) writeDebugInfo(w io.Writer) {
 		fmt.Fprintf(w, "%s\n", info)
 	}
 }
+
+func (o *ownerImpl) markProcessorDown(ctx context.Context,
+	p *model.ProcessorInfo) error {
+	statuses, err := o.cfRWriter.GetAllTaskStatus(ctx, p.ChangeFeedID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	positions, err := o.cfRWriter.GetAllTaskPositions(ctx, p.ChangeFeedID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// lookup the task position for the processor
+	pos, exist := positions[p.CaptureID]
+	if !exist {
+		log.Warn("unkown processor deletion detected",
+			zap.String("processorid", p.ID),
+			zap.String("captureid", p.CaptureID))
+		return nil
+	}
+	// lookup the task position for the processor
+	status, exist := statuses[p.CaptureID]
+	if !exist {
+		log.Warn("unkown processor deletion detected",
+			zap.String("processorid", p.ID),
+			zap.String("capture", p.CaptureID))
+		return nil
+	}
+	snap := status.Snapshot(p.ChangeFeedID,
+		p.CaptureID,
+		pos.CheckPointTs)
+	log.Info("mark processor down",
+		zap.String("processorid", p.ID),
+		zap.String("captureid", p.CaptureID),
+		zap.String("changefeed", p.ChangeFeedID),
+		zap.Reflect("tables", snap.Tables))
+	o.processorLock.Lock()
+	o.markDownProcessor = append(o.markDownProcessor, snap)
+	delete(o.activeProcessors, p.ID)
+	o.processorLock.Unlock()
+	return nil
+}
+
+func (o *ownerImpl) markProcessorActive(ctx context.Context,
+	p *model.ProcessorInfo) error {
+	o.processorLock.Lock()
+	o.activeProcessors[p.ID] = p
+	o.processorLock.Unlock()
+	return nil
+}
+
+func (o *ownerImpl) rebuildProcessorEvents(ctx context.Context,
+	processors []*model.ProcessorInfo) error {
+	current := make(map[string]*model.ProcessorInfo)
+	for _, p := range processors {
+		current[p.ID] = p
+		if _, ok := o.activeProcessors[p.ID]; !ok {
+			if err := o.markProcessorActive(ctx, p); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	for _, p := range o.activeProcessors {
+		if _, ok := current[p.ID]; !ok {
+			if err := o.markProcessorDown(ctx, p); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (o *ownerImpl) watchProcessorInfo(ctx context.Context) error {
+	ctx = clientv3.WithRequireLeader(ctx)
+
+	rev, processors, err := o.etcdClient.GetAllProcessors(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// before watching, rebuild events according to
+	// the existed processors. This is necessary because
+	// the etcd events may be compacted.
+	if err := o.rebuildProcessorEvents(ctx, processors); err != nil {
```

Comment:
Consider the scenario
1. we have two captures
2. the owner capture crashes (kill -9) and the second capture becomes the owner
3. the processor info of the original owner will be deleted because of session expired
4. The second capture takes a longer time to be an owner than the processor info expired, in the current logic, the new owner doesn't aware of the crashed processor when it `rebuildProcessorEvents`.

Reply:
Yes, the events in that scenario would not be detected, but it is OK here. The case has been covered in the `cleanUpStaleTasks` function. It checks the tasks with the active processors, and if the processor associated with the task is not active, the task will be rebalanced. When an owner starts to be in charge, the processors that have expired would not be loaded as an active processor, the processors that being loaded would not lose the expiration events.

Reply:
Oh yes you are right. I find the bug does not locate here but in the `CaptureInfoWatcher`.

Still in this scenario, after the second capture becomes owner, it rebalances tables based on the `capture list` in `o.captures`, but the capture info of the crashed capture is not deleted, so some tables are rebalanced to the crashed capture.

The implementation of `CaptureInfoWatcher` doesn't consider the owner role very well, it should also have some similar logic like `ProcessorInfoWatcher`. Besides these two watchers may do some conflict works, should we combine these two watchers or leave them alone to do different jobs, what's your suggestion

Reply:
There are two problems with the capture watcher, the first one as you said, the capture info is not deleted. The second one is all the roles including the non-owner captures are watching the events, this may handle the capture deletion multiple times.

I am planning to solve these problems by refactoring the capture:
* Use one session for the capture self, its election, and all its processors.
* Remove the capture info watcher(I am not sure is it safe to remove it)

Reply:
Sounds great! we can refine this in some more PRs.

Reply:
Yes, agree with you.